### ISSUE TITLE
[Suggestions] Cave/oregen shape ideas

### ISSUE THREAD

**Issue by superfluke** [PersonTheCat/CaveGenerator#13]
Heya, ran across some interesting shapes working on something else and thought they could be morphed into some interesting caves. I know you were interested in other cave ideas at one point, so thought I'd share them here in case you found them useful. Would probably take some tinkering with extra domain warping or combining with other noise functions, but might be a cool base shape for cave or even as-is for ore gen. I use the FastNoise library for the code snippets. 

Gradient Perturbed value noise:
Keeps the kind of blocky shape of value noise which turns into conneced rooms. Increasing the amplitude/frequency of the perturbnoise will distort the blocky shape even more. 
![image](https://user-images.githubusercontent.com/20022336/44739583-2bce7f80-aac6-11e8-8af4-399494811695.png)
![2018-08-28_13 07 14](https://user-images.githubusercontent.com/20022336/44739833-e52d5500-aac6-11e8-99c9-7b7c3ca73d61.png)
```
init(){
	FastNoise perturbnoise = new FastNoise();
	FastNoise valnoise = new FastNoise();
	perturbnoise.SetFrequency(0.035f);
	valnoise.SetNoiseType(NoiseType.Value);
	valnoise.SetFrequency(0.06F);
}

genTerain(){
                for (int y = 0; y <= height; y++)  {
                	Vector3f pvec = new Vector3f(realX, y, realZ);
        		perturbnoise.GradientPerturbFractal(pvec);
        		if(valnoise.GetNoise(pvec.x, pvec.y*2, pvec.z) > 2.2)
        			primer.setBlockState(x, y, z, Blocks.STONE.getDefaultState());
                }
}
```

Cellular noise lookup:
Football shaped segments connected together create a kind of winding cave system. Increasing the Jitter of the cavenoise would vary the cell size/shape a bit more as well. 
![image](https://user-images.githubusercontent.com/20022336/44739464-d1352380-aac5-11e8-865c-533a38c93e62.png)
![2018-08-28_13 10 43](https://user-images.githubusercontent.com/20022336/44740244-f0cd4b80-aac7-11e8-99bf-40375b33cba7.png)
```
init(){
	FastNoise perlin = new FastNoise();
	FastNoise cavenoise = new FastNoise();
	cavenoise.SetNoiseType(NoiseType.Cellular);
	cavenoise.SetFrequency(0.1F);
	cavenoise.SetCellularNoiseLookup(perlin);
	cavenoise.SetCellularReturnType(CellularReturnType.NoiseLookup);
}

genTerain(){
		for (int y = 0; y <= height; y++) {
                	float cave = cavenoise.GetNoise(realX*25, y*65, realZ*25); //could probably just increase cavenoise frequency instead but was playing with values here 
            		if(cave < 0.0F){
            			primer.setBlockState(x, y, z, Blocks.STONE.getDefaultState());
            		}
                	
                }
}
```

**Comment by PersonTheCat** [comment]
Hey, this is a very detailed report and I think it could be very feasible to implement these features. I've looked at FastNoise before and have even taken inspiration from some things like their hash generator, but I for some reason decided against using too much of their work. I guess there's no reason not to, though, seeing as it's all available in the public domain. I'll start looking into what kinds of shapes I can produce and what the parameters might look like in the cave configs sometime later this week. I'm a little busy with classes, so I'm not exactly sure how long that'll take. 

Thanks a lot for the suggestion!

**Comment by PersonTheCat** [comment]
Finally making the switch to FastNoise for everything now. Users can now optionally specify the the NoiseType parameter anywhere noise values are used. I'll leave this open until the rewrite is finished, but a testable build will be available here in the next hour or so.

**Comment by PersonTheCat** [comment]
Released in version 0.13. Still working on basically letting users choose whichever FastNoise arguments they want, but I'll go ahead and close this issue.
